go/vt/tabletserver/query_executor.go

```diff
@@ -106,8 +106,46 @@ func (qre *QueryExecutor) Execute() (reply *mproto.QueryResult) {
 			conn := qre.getConn(qre.qe.connPool)
 			defer conn.Recycle()
 			reply = qre.execSQL(conn, qre.query, true)
+		}
+		if reply != nil {
+			return reply
+		}
+		// autocommit for single dml
+		transactionID := qre.qe.txPool.Begin(qre.ctx)
+		qre.logStats.AddRewrittenSql("begin", time.Now())
+		isQueryFailed := false
+		defer func() {
+			if !isQueryFailed {
+				qre.qe.Commit(qre.ctx, qre.logStats, transactionID)
+				qre.logStats.AddRewrittenSql("commit", time.Now())
+			}
```

Comment:
The current scheme is that the exec functions panic. So, you have to catch them in a defer block and rollback. After rollback, you have to re-panic for handleExecError to catch it and return the rpc error.
We do need to clean this up by switching to using error returns.
Even then, we may still have to catch unexpected panics so we can rollback.